launcher/modplatform/atlauncher/ATLPackInstallTask.cpp

```diff
@@ -337,58 +335,33 @@ QString PackInstallTask::getDirForModType(ModType type, QString raw)
-            emitFailed(tr("Failed to get local metadata index for %1").arg(uid));
-            return Q_NULLPTR;
-        }
-
-        vlist->waitToLoad();
-
-        if (m_version.loader.recommended || m_version.loader.latest) {
-            for (int i = 0; i < vlist->versions().size(); i++) {
-                auto version = vlist->versions().at(i);
-                auto reqs = version->requiredSet();
-
-                // filter by minecraft version, if the loader depends on a certain version.
-                // not all mod loaders depend on a given Minecraft version, so we won't do this
-                // filtering for those loaders.
-                if (m_version.loader.type != "fabric") {
-                    auto iter = std::find_if(reqs.begin(), reqs.end(), [](const Meta::Require& req) { return req.uid == "net.minecraft"; });
-                    if (iter == reqs.end())
-                        continue;
-                    if (iter->equalsVersion != m_version.minecraft)
-                        continue;
-                }
-
-                if (m_version.loader.recommended) {
-                    // first recommended build we find, we use.
-                    if (!version->isRecommended())
-                        continue;
-                }
-
-                return version->descriptor();
-            }
-
-            emitFailed(tr("Failed to find version for %1 loader").arg(m_version.loader.type));
-            return Q_NULLPTR;
-        } else if (m_version.loader.choose) {
-            // Fabric Loader doesn't depend on a given Minecraft version.
-            if (m_version.loader.type == "fabric") {
-                return m_support->chooseVersion(vlist, Q_NULLPTR);
-            }
-
-            return m_support->chooseVersion(vlist, m_version.minecraft);
-        }
+    QString loaderVersion;
+    if (m_version.loader.recommended) {
+        loaderVersion = "recommended";
+    } else if (m_version.loader.latest) {
+        loaderVersion = "latest";
+    } else if (m_version.loader.choose) {
+        loaderVersion = "choose";
+    } else {
+        loaderVersion = m_version.loader.version;
     }
-
-    if (m_version.loader.version == Q_NULLPTR || m_version.loader.version.isEmpty()) {
-        emitFailed(tr("No loader version set for modpack!"));
+    QString err;
+    auto version = APPLICATION->metadataIndex()->getVersionForLoader(uid, m_version.loader.type, loaderVersion, m_version.minecraft, err);
+    if (!err.isEmpty()) {
+        emitFailed(err);
         return Q_NULLPTR;
     }
 
-    return m_version.loader.version;
+    if (m_version.loader.choose) {
+        auto vlist = APPLICATION->metadataIndex()->get(uid);
+        // Fabric Loader doesn't depend on a given Minecraft version.
+        if (m_version.loader.type == "fabric" || m_version.loader.type == "quilt") {
+            return m_support->chooseVersion(vlist, Q_NULLPTR);
+        }
+
+        return m_support->chooseVersion(vlist, m_version.minecraft);
+    }
```

Comment:
Also need to remove duplicated logic above

```suggestion
    if (m_version.loader.choose) {
        auto vlist = APPLICATION->metadataIndex()->get(uid);
        // Fabric Loader doesn't depend on a given Minecraft version.
        if (m_version.loader.type == "fabric" || m_version.loader.type == "quilt") {
            return m_support->chooseVersion(vlist, Q_NULLPTR);
        }

        return m_support->chooseVersion(vlist, m_version.minecraft);
    } else {
        QString err;
        auto version = APPLICATION->metadataIndex()->getVersionForLoader(uid, m_version.loader.type, loaderVersion, m_version.minecraft, err);
        if (!err.isEmpty()) {
            emitFailed(err);
            return Q_NULLPTR;
        }
    }
```

Reply:
Now it makes sense to choose before the `getVersionForLoader`
It should be fixed now